File: src/15.三数之和.ts
/*
 * @lc app=leetcode.cn id=15 lang=typescript
 *
 * [15] 三数之和
 */

// @lc code=start
function threeSum(nums: number[]): number[][] {
	const res: Array<Array<number>> = [];

	// sort
	nums.sort((a,b)=>a-b)

	const len:number=nums.length
	const end:number=len-2

	// boundary
	if (nums[0] <= 0 && nums[len - 1] >= 0) {
		for (let i:number = 0; i < end;) {
			// 取负数
			if(nums[i]>0) break
			let head:number = i+1
			let tail:number = len-1
			while(head<tail){
				let first:number=nums[head]
				let second:number=nums[tail]
				const result:number = nums[i]+first+second
				if(result == 0) {
					res.push([nums[i], first, second])
				}
				if(result<=0){
					while(nums[head]==nums[++head]){}
				}else if(result>0){
					while(nums[tail]==nums[--tail]){}
				}else{
					break
				}
			}
			while(nums[i]==nums[++i]){}
		}
	}
	// 双指针

	return res;
}
// @lc code=end
// threeSum([-1, 0, 1, 2, -1, -4])

File: src/15.三数之和.go
/*
 * @lc app=leetcode.cn id=15 lang=golang
 *
 * [15] 三数之和
 */
package leetcode

import "sort"

// @lc code=start
func threeSum(nums []int) (ans [][]int) {
	n := len(nums)

	// sort
	sort.Ints(nums)

	for first := 0; first < n; first++ {
		// jump 
		if first > 0 && nums[first] == nums[first - 1] {
			continue
		}

		third := n - 1
		target := -1 * nums[first]
		for second := first + 1; second < n; second++ {
			if second > first + 1 && nums[second] == nums[second - 1] {
				continue
			}

			for second < third && nums[second] + nums[third] > target {
				third --
			}
			
			result := nums[first] + nums[second] + nums[third]

			if second == third {
				break
			}
			
			if result == 0 {
				ans = append(ans, []int{nums[first], nums[second], nums[third]})
			}
		}
	}


	return
}
// @lc code=end

